src/ipc/ipc3/handler.c

```diff
@@ -386,6 +386,13 @@ static int ipc_stream_pcm_free(uint32_t header)
 		return -EINVAL;
 	}
 
+	/* pipeline_reset() crashes if source_comp is NULL */
+	if (!pcm_dev->cd->pipeline->source_comp) {
+		ipc_cmd_err(&ipc_tr, "ipc: comp %d source comp not found",
+			    free_req.comp_id);
+		return -EINVAL;
+	}
+
```

Comment:
Could we have the check in pipeline_comp_reset() and make pipeline_reset() safe against condition of partially set up pipeline (with source_comp as NULL). This would fix a much larger set of cases (including IPC4) against this possibility.

Reply:
Sure. There just is no evidence that IPC4 would suffer from this kind of problem. But if you think that is better approach, then I can send my first fix. I decided against it and came up with this version, since I tried to avoid affecting IPC4-code, as the problem was found from IPC3.